### ISSUE TITLE
Feature: Human-in-the-Loop Review + Active Learning (feedback store → retraining hooks)

### ISSUE THREAD

**Issue by hoangsonww** [hoangsonww/Spot-the-Scam-AI-Job-Fraud#2]
### Summary

Add a production-grade feedback loop so reviewers can **confirm/override** model decisions in the UI, persist those judgements, and feed them back into the training pipeline for **active learning** (uncertainty sampling + periodic recalibration/retraining). This closes the loop from prediction → human decision → improved model.

---

### Why

* Repo already supports **gray-zone** and calibrated scores; perfect foundation for triage.
* Without labeled feedback from reviewers, model quality plateaus and calibration drifts.
* Lightweight storage + endpoints + small UI lift unlocks **continuous improvement**.

---

### Scope (MVP)

1. **Feedback storage (append-only)**

   * Write once, immutable rows with: `id, created_at, request_id, text_hash, features_hash, model_version, proba, predicted_label, reviewer_label, rationale(optional), notes(optional), source(split), flags`.
   * Storage: start with **Parquet** under `tracking/feedback/` (no DB required); later swappable to Postgres.
2. **FastAPI endpoints**

   * `POST /feedback` (create one or batch)
   * `GET /feedback?since=...&limit=...` (for dashboards)
   * `GET /cases?policy=gray-zone&limit=...` (serve uncertain cases to UI)
   * Pydantic schemas + simple rate-limit + PII redaction guard (e.g., emails, phones → masked in logs).
3. **UI – Triage drawer**

   * In the Next.js dashboard, add a **“Review Queue”**:

     * Pull `/cases?policy=gray-zone` (e.g., `p in [0.4,0.6]`) or top-K by entropy.
     * Show model rationale card + tokens/SHAP already surfaced.
     * Buttons: **Confirm Legit** / **Confirm Fraud** / **Unsure** + notes.
     * On submit → `POST /feedback`.
4. **Training pipeline consumption**

   * New step in training: merge canonical training data with `tracking/feedback/*.parquet`.
   * Deduplicate by `(text_hash, model_version, reviewer_label)`; prefer latest reviewer label.
   * Emit **slice metrics** “with” and “without” feedback to show lift.
5. **Active learning sampler**

   * Script `scripts/sample_uncertain.py`:

     * Reads recent predictions, selects top-N most uncertain or with largest calibration error by slice, writes `experiments/tables/active_sample.csv`.
   * UI button **“Load N more for review”** calls that sampler (or hits `/cases`).
6. **Calibration & gating**

   * After retrain, compute Brier/reliability; if **ECE/WCE improved** and PR-AUC not worse by >δ, write a `gates.json` “PASS”.
   * Persist plots under `experiments/figs/calibration_after_feedback/*`.

---

### Acceptance Criteria

* [ ] `POST /feedback` persists Parquet rows with required fields; basic schema validation & redaction.
* [ ] Frontend allows a reviewer to submit labels for queued cases; optimistic UI + error toasts.
* [ ] Training pipeline can be run with `USE_FEEDBACK=1` to incorporate labels and outputs side-by-side metrics (`experiments/tables/metrics_with_feedback.csv`).
* [ ] New `make` targets:

  * `make serve-queue` (runs API + exposes `/cases`)
  * `make review-sample` (runs sampler to refresh queue artifacts)
  * `make retrain-with-feedback`
* [ ] Documentation updated (`INSTRUCTIONS.md` + `README.md` “Human-in-the-Loop”).
* [ ] CI green (lint, tests).

---

### API Sketch (FastAPI)

```py
# src/api/schemas.py
class FeedbackIn(BaseModel):
    request_id: str
    model_version: str
    proba: float
    predicted_label: Literal["fraud","legit"]
    reviewer_label: Literal["fraud","legit","unsure"]
    text_hash: str
    features_hash: str
    rationale: str | None = None
    notes: str | None = None

# src/api/routes_feedback.py
@router.post("/feedback", status_code=201)
def post_feedback(items: list[FeedbackIn]):
    # redact PII in notes/rationale before writing
    # append to tracking/feedback/feedback-YYYYMMDD.parquet (partitioned)
    ...
```

---

### Frontend Sketch (Next.js)

* New page `/review`:

  * `GET /cases?policy=gray-zone&limit=50`
  * Card shows: title snippet, score, rationale, top tokens, latency.
  * Actions → `POST /feedback`.
* Add a small **counter** badge in the navbar “Review (17)”.

---

### Training Integration

* In `src/train/ingest.py`:

  * `feedback = read_parquet("tracking/feedback/*.parquet") if USE_FEEDBACK`
  * Join on `text_hash` (or `request_id` if you persist source text separately under hashed filenames).
  * Prefer reviewer label as ground truth where available.
* Output diff table:

  * `experiments/tables/metrics_delta_after_feedback.csv`
  * Columns: metric, base, with_feedback, delta.

---

### Active Learning Policies (configurable)

* **Uncertainty:** entropy(p), margin(|p−0.5|).
* **Calibration error:** abs(p − empirical_rate) by token slice.
* **Diversity guard:** k-means on TF-IDF to avoid near-duplicates in same batch.

---

### Security/Privacy Guards

* PII redaction regex for emails/phones/SSNs in rationale/notes.
* Cap payload size; 429 on flood.
* Feedback files are **append-only**; mutability requires a separate “corrections” log.

---

### Out of Scope (future issues)

* Reviewer authentication & roles.
* Postgres storage + audit trails.
* Auto-promotion to MLflow Registry on “gates PASS”.
* True online learning.

---

### Bite-sized sub-tasks (grab-and-go)

1. **Backend:** add Parquet writer + partitioning by date.
2. **Backend:** implement `/feedback` with Pydantic validation + simple redaction.
3. **Backend:** implement `/cases?policy=gray-zone` using saved prediction dumps.
4. **Frontend:** build Review queue UI + optimistic submit.
5. **ML:** merge feedback in training; write delta metrics table.
6. **Docs:** add “Human-in-the-Loop” section with diagrams and `make` recipes.
